downstairs/src/extent_inner_raw.rs

```diff
@@ -0,0 +1,2375 @@
+            out.push(ctx.map(|c| DownstairsBlockContext {
+                block: block_start + i as u64,
+                block_context: c.block_context,
+                on_disk_hash: c.on_disk_hash,
+            }));
+        }
+        self.buf.set(buf);
+        Ok(out)
+    }
+
+    /// Write out the active context array and metadata section of the file
+    ///
+    /// This is done in a single write, so it should be atomic.
+    ///
+    /// # Panics
+    /// `active_context.len()` must match `self.block_count()`, and the function
+    /// will panic otherwise.
+    fn write_active_context_and_metadata(
+        &self,
+        file: &File,
+        active_context: &[ContextSlot],
+        dirty: bool,
+        flush_number: u64,
+        gen_number: u64,
+    ) -> Result<(), CrucibleError> {
+        assert_eq!(active_context.len(), self.block_count() as usize);
+
+        // Serialize bitpacked active slot values
+        let mut buf = self.buf.take();
+        buf.clear();
+        for c in active_context.chunks(8) {
+            let mut v = 0;
+            for (i, slot) in c.iter().enumerate() {
+                v |= (*slot as u8) << i;
+            }
+            buf.push(v);
+        }
+
+        let d = OnDiskMeta {
+            dirty,
+            flush_number,
+            gen_number,
+            ext_version: EXTENT_META_RAW,
+        };
+        let mut meta = [0u8; BLOCK_META_SIZE_BYTES as usize];
+        bincode::serialize_into(meta.as_mut_slice(), &d).unwrap();
+        buf.extend(meta);
+
+        let offset = self.active_context_offset();
+
+        nix::sys::uio::pwrite(file.as_raw_fd(), &buf, offset as i64).map_err(
+            |e| CrucibleError::IoError(format!("writing metadata failed: {e}")),
+        )?;
+        self.buf.set(buf);
+
+        Ok(())
+    }
+
+    /// Decodes the active contexts from the given file
+    ///
+    /// The file descriptor offset is not changed by this function
+    fn get_active_contexts(
+        &self,
+        file: &File,
+    ) -> Result<Vec<ContextSlot>, CrucibleError> {
+        let mut buf = self.buf.take();
+        buf.resize(self.active_context_size() as usize, 0u8);
+        let offset = self.active_context_offset();
+        nix::sys::uio::pread(file.as_raw_fd(), &mut buf, offset as i64)
+            .map_err(|e| {
+                CrucibleError::IoError(format!(
+                    "could not read active contexts: {e}"
+                ))
+            })?;
+
+        let mut active_context = vec![];
+        for bit in buf
+            .iter()
+            .flat_map(|b| (0..8).map(move |i| b & (1 << i) == 0))
+            .take(self.block_count() as usize)
+        {
+            // Unpack bits from each byte
+            active_context.push(if bit {
+                ContextSlot::A
+            } else {
+                ContextSlot::B
+            });
+        }
+        assert_eq!(active_context.len(), self.block_count() as usize);
+        Ok(active_context)
```

Comment:
You did not `self.buf.set(buf);` here. Is that on purpose?

Reply:
Nope, fixed in a8b3705f33851d1c4a6d66916b066f0d5e299b02